geo/geo.go

```diff
@@ -0,0 +1,243 @@
+package geo
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+type GeoProvider interface {
+	GetLocation(ip string) (*Location, error)
+	FileName() string
+	DownloadUrl() string
+	LastUpdate() time.Time
+	NeedUpdate() bool
+	MaybeUpdate() error
+	Ready() bool
+	SetReady(bool)
+	SetFileName(string)
+}
+
+type Location struct {
+	IP          string `json:"ip"`
+	CountryCode string `json:"countryCode"`
+	Country     string `json:"country"`
+	City        string `json:"city"`
+}
+
+var emptyLocation = Location{}
+
+func (l *Location) Empty() bool {
+	if l == nil {
+		return true
+	}
+	return *l == emptyLocation
+}
+
+func (l *Location) Match(location *Location) bool {
+	if len(l.IP) > 0 && location.IP == l.IP {
+		return true
+	}
+	if len(l.CountryCode) > 0 && strings.ToLower(location.CountryCode) == strings.ToLower(l.CountryCode) {
+		return true
+	}
+	return false
+}
+
+type IPFilter struct {
+	DBPath    string        `json:"dbpath"`
+	Download  bool          `json:"download"`
+	Allow     []Location    `json:"allow"`
+	Disallow  []Location    `json:"disallow"`
+	Providers []GeoProvider `json:"-"`
+}
+
+func (f *IPFilter) Empty() bool {
+	if f == nil {
+		return true
+	}
+	for _, loc := range f.Allow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	for _, loc := range f.Disallow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (f *IPFilter) NeedGeoInfo() bool {
+	if f.Empty() {
+		return false
+	}
+	for _, loc := range f.Allow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	for _, loc := range f.Disallow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (f *IPFilter) AllowIP(ip string) (bool, error) {
+	if f.Empty() {
+		return true, nil
+	}
+
+	var loc *Location
+	if f.NeedGeoInfo() {
+		loc = f.GetLocation(ip)
+	} else {
+		loc = &Location{IP: ip}
+	}
+
+	return f.AllowLocation(loc), nil
+}
+
+func (f *IPFilter) GetLocation(ip string) *Location {
+	for _, p := range f.Providers {
+		if p.Ready() {
+			loc := getLocationFromProvider(ip, p.(GeoProvider))
+			if !loc.Empty() {
+				return &loc
+			}
+		}
+	}
+	return &Location{CountryCode: "UNKNOWN", IP: ip}
+}
+
+func (f *IPFilter) AllowLocation(loc *Location) bool {
+	if loc.Empty() {
+		return true
+	}
+
+	for _, l := range f.Disallow {
+		if l.Match(loc) {
+			return false
+		}
+	}
+
+	empty := true
+	for _, l := range f.Allow {
+		if l.Match(loc) {
+			return true
+		}
+		if !l.Empty() {
+			empty = false
+		}
+	}
+
+	return empty
+}
+
+func getLocationFromProvider(ip string, p GeoProvider) Location {
+	loc, err := p.GetLocation(ip)
+	if err != nil {
+		log.Println(err)
+	}
+	return *loc
+}
+
+func GetJsonFile(url, filename string) ([]byte, error) {
+	if s, err := os.Stat(filename); os.IsNotExist(err) || s.Size() == 0 {
+		f, err := os.Create(filename)
```

Comment:
You need to remove the file for ANY error that occurs before L177, or you should create the file just before writing to file